checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java

```diff
@@ -429,61 +679,70 @@ private void trackInvocationResult(Set<Obligation> obligations, Node node) {
             FluentIterable.from(obligationContainingMustCallAlias.resourceAliases)
                 .append(tmpVarAsResourceAlias)
                 .toSet();
         obligations.remove(obligationContainingMustCallAlias);
         obligations.add(new Obligation(newResourceAliasSet));
       }
+      // It is not an error if there is no Obligation containing the must-call alias. In that
+      // case, what has usually happened is that no Obligation was created in the first place.
+      // For example, when checking the invocation of a "wrapper stream" constructor, if the
+      // argument in the must-call alias position is some stream with no must-call obligations like
+      // a
+      // ByteArrayInputStream, then no Obligation object will have been created for it and therefore
+      // obligationContainingMustCallAlias will be null.
     } else {
-      // If mustCallAlias is neither a field nor a local already in the set of obligations,
-      // add it to a new set.
+      // If there is no MustCallAlias, create a new Obligation for the resouce alias.
+      ResourceAlias tmpVarAsResourceAlias = new ResourceAlias(new LocalVariable(tmpVar), tree);
       obligations.add(new Obligation(ImmutableSet.of(tmpVarAsResourceAlias)));
     }
   }
 
   /**
    * Determines if the result of the given method or constructor invocation node should be tracked
    * in {@code obligations}. In some cases, there is no need to track the result because the
-   * obligations are already satisfied in some other way or there cannot possibly be obligations
-   * because of the structure of the code.
+   * must-call obligations are already satisfied in some other way or there cannot possibly be
+   * must-call obligations because of the structure of the code.
    *
-   * <p>Specifically, an invocation result does not need to be tracked if any of the following are
-   * true: (1) the invocation is a call to a {@code this()} or {@code super()} constructor, (2) if
-   * the method's return type (or, the constructor result, if this is an invocation of a
-   * constructor) is annotated with MustCallAlias and the argument passed in this invocation in the
-   * corresponding position is an owning field, or (3) if the method's return type (or the
-   * constructor result) is non-owning, which can either be because the method has no return type or
-   * because it is annotated with {@link NotOwning}.
+   * <p>Specifically, an invocation result does NOT need to be tracked if any of the following is
+   * true:
    *
-   * <p>This method can also side-effect obligations, if node is a super or this constructor call
-   * with MustCallAlias annotations, by removing that obligation.
+   * <ul>
+   *   <li>The invocation is a call to a {@code this()} or {@code super()} constructor.
+   *   <li>The method's return type is annotated with MustCallAlias and the argument passed in this
+   *       invocation in the corresponding position is an owning field.
```

Comment:
Why does this have to be an owning field as opposed to some other expression that is owning?

Reply:
Because for other owning expressions, such as local variables or method parameters, the MustCallAlias result is added to the resource alias set in the Obligation for the owning expression.

Owning fields are treated specially because they are handled separately. If the result of a method invocation is MustCallAlias with an owning field, then there is no need for this algorithm to track it, because it is guaranteed to be closed via the owning field (that is the definition of an owning field, after all).

Reply:
Do you think this logic needs to be laid out in the source code?